{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MICROBIOME 16S PIPELINE ION TORRENT.\n",
    "\n",
    "# This pipeline was developed for microbiome 16S data analysis comming from Ion Torrent technology (Ion 16S™ Metagenomics Kit) in the Hematological Tumor Unit of\n",
    "# Universidad Complutense de Madrid/Hospital Doce de Octubre/CNIO.\n",
    "\n",
    "# BEFORE STARTING:\n",
    "\n",
    "# A. Ion 16S™ Metagenomics Kit uses 2 pools of primers (Pool 1: V2, V4, V8 and Pool 2: V3, V6-7 and V9) and it generates a FASTQ file per sample which contains sequences \n",
    "# comming from all V regions. This kit uses bidireccional sequencing (not paired-end) with forward (F) and reverse (R) primers, so both sequence types are included in \n",
    "# each FASTQ file. In addition, sequences are demultiplexed and molecular barcodes are removed by the sequence machine. PRIMERS SEQUENCES ARE PROPIETARY AND UNKNOWN!!!\n",
    "\n",
    "# B. CURRENT LIMITATIONS Ion 16S™ Metagenomics Kit: Although the Ion Torrent software used to analyse data (Ion Reporter) adds up OTUs comming from the different V regions,\n",
    "# this process has important bias and it is not considered as suitable by a high number of Qiime2 and Ion Torrent users. The key is that V primers don't overlap between them\n",
    "# so this data can't be treated as other omics data (e.g RNA-seq counts) which are assembled before counting. In this case, if OTUs from different V regions are added up, we\n",
    "# will overestimate the richness in our community because some taxa will be count double, triple or more!!! IF THIS APPROACH IS FOLLOWED, THIS ASPECT SHOULD BE MENTIONED IN \n",
    "# THE DISCUSSION SPECIFYING THAT THERE IS NOT A SUITABLE APPROACH FOR MERGING ACCROSS V REGIONS AT DATE OF MARCH 2023. \n",
    "\n",
    "# C. To cover the great disadvantage of this kit, this pipeline will generate outputs for all individual V region and for a \"consensus\" (created as Ion Reporter)\n",
    "\n",
    "# THIS CODE WAS USED TO REORGANIZE FASTQ FILES BY V REGION (PREVIOUSLY DIVIDED BY SAMPLE AND V REGION), GENERATE COUNTS STATISTICS FOR EACH V REGION AND \n",
    "# WRITE A MANIFEST FILE TO IMPORT FASTQ FILES FOR EACH V REGION IN QIIME2.\n",
    "\n",
    "# This code was generated by Katherine Maki and partners (add citation in the paper)\n",
    "import os\n",
    "import shutil\n",
    "import gzip\n",
    "import pandas as pd # version 1.4.0\n",
    "\n",
    "### move or copy fastq or fastq.gz files to a new directory structure\n",
    "def reorg_metagenomics_files(oldPath='./cutprimers/cutprimers_output',\n",
    "                             newPath='./cutprimers/cutprimers_Vsorted'):\n",
    "    if not os.path.isdir(newPath):\n",
    "        os.mkdir(newPath)\n",
    "        \n",
    "    ## Read file tree in oldPath and get names of fastq files\n",
    "    def getFileInfo(oldPath,newPath):\n",
    "        filesOld,filesNew = [],[]\n",
    "        v_regions = set()   \n",
    "        for root, dirs, filenames in os.walk(oldPath):\n",
    "            for file in filenames:\n",
    "                if file[-6:].lower() == '.fastq' or file[-9:].lower() == '.fastq.gz':\n",
    "                    filesOld.append(os.path.join(root,file))\n",
    "                    f = file.replace('FR.','')\n",
    "                    v = f.split('.')[1]\n",
    "                    v_regions.add(v)\n",
    "                    fNew = f.replace('.' + v,'_' + v)\n",
    "                    filesNew.append(os.path.join(newPath,v,fNew))\n",
    "        return filesOld,filesNew,v_regions\n",
    "    filesOld,filesNew,v_regions = getFileInfo(oldPath,newPath)\n",
    "    assert len(filesOld), 'no files found in \"{}\"'.format(oldPath)\n",
    "    \n",
    "    ## prompt for user input before manipulating files\n",
    "    def getUserInputs(v_regions):     \n",
    "        print('\\nSubdirectories:')\n",
    "        for v in sorted(v_regions):\n",
    "            print('\\t{}'.format(v))\n",
    "            \n",
    "        inputs = {}\n",
    "        def getInput(dictionary,key,msg):\n",
    "            while dictionary.get(key) not in {'y','n'}:\n",
    "                dictionary[key] = input(msg)\n",
    "        getInput(inputs,'copy','Copy fastq files into the above subdirectories? [y/n] ')\n",
    "        if inputs['copy'] == 'n':\n",
    "            getInput(inputs,'move','Move fastq files into the above subdirectories? [y/n] ')\n",
    "        else:\n",
    "            getInput(inputs,'gzip','Compress (gzip) the new fastq files? [y/n] ')\n",
    "        print()\n",
    "        return inputs\n",
    "    inputs = getUserInputs(v_regions)\n",
    "        \n",
    "    ## if compressing files, then add any missing .gz extension to names of new files\n",
    "    def gzipFileNames(files):\n",
    "        return [f  + '.gz' if f[-3:] != '.gz' else f for f in files]\n",
    "    if inputs.get('gzip') == 'y':\n",
    "        filesNew = gzipFileNames(filesNew)\n",
    "    \n",
    "    ## copy fastq files, optionally use gzip to compress any uncompressed files\n",
    "    def copyFiles(filesOld,filesNew,compress='n'):\n",
    "        if compress == 'n':\n",
    "            print('\\t{} files total, copying file     '.format(len(filesNew)),end='')\n",
    "            for i,(fOld,fNew) in enumerate(zip(filesOld,filesNew)):\n",
    "                print('\\b\\b\\b\\b{0:4d}'.format(i+1),end='')\n",
    "                os.makedirs(os.path.dirname(fNew),exist_ok=True)\n",
    "                shutil.copy2(fOld,fNew)\n",
    "            print('\\n\\tAll fastq files copied!\\n')\n",
    "        else:\n",
    "            print('\\t{} files total, compressing file    '.format(len(filesNew)),end='')\n",
    "            for i,(fOld,fNew) in enumerate(zip(filesOld,filesNew)):\n",
    "                print('\\b\\b\\b\\b{0:4d}'.format(i+1),end='')\n",
    "                os.makedirs(os.path.dirname(fNew),exist_ok=True)\n",
    "                if fNew[-3:] == '.gz' and fOld[-3:] != '.gz':\n",
    "                    with open(fOld,'rb') as f:\n",
    "                        with gzip.open(fNew,'wb') as fzip:\n",
    "                            shutil.copyfileobj(f,fzip)\n",
    "                else:\n",
    "                     shutil.copy2(fOld,fNew)\n",
    "            print('\\n\\tAll fastq files copied as .gz files!\\n')\n",
    "            \n",
    "    ## move files to new location -- obviously faster than copying\n",
    "    def moveFiles(filesOld,filesNew):\n",
    "        for fOld,fNew in zip(filesOld,filesNew):\n",
    "            os.makedirs(os.path.dirname(fNew),exist_ok=True)\n",
    "            shutil.move(fOld,fNew)\n",
    "            \n",
    "    ## execute move or copy        \n",
    "    if inputs.get('move') == 'y':\n",
    "        moveFiles(filesOld,filesNew)\n",
    "    elif inputs['copy'] == 'y':\n",
    "        copyFiles(filesOld,filesNew,inputs['gzip'])\n",
    "    \n",
    "    ## write aggregate counts file\n",
    "    def aggregate_counts_data(writeFile,oldPath,stats_rows=True):\n",
    "        # Aggregates total counts from \"counts.txt\" files in the mainPath directory \n",
    "        # tree. File has a row for each sample and a column for each primer.\n",
    "        # If stats_rows is True, then it will add rows of statistics to the end.\n",
    "        \n",
    "        dfs = []\n",
    "        for root, dirs, filenames in os.walk(oldPath):\n",
    "            for file in filenames:\n",
    "                if file == 'counts.txt':\n",
    "                    dfs.append(pd.read_csv(os.path.join(root,file),sep='\\t',\n",
    "                                           names=['name','count']))\n",
    "        df = pd.concat(dfs)\n",
    "        df = df[df['name'] != 'total']\n",
    "        df = pd.concat((df.name.str.split('.',expand=True),df['count']),axis=1)\n",
    "        df = df.pivot_table(index=0,columns=1,values='count')\n",
    "        if stats_rows:\n",
    "            df = pd.concat((df,df.agg(['mean','std','sem'])))\n",
    "        df.to_csv(writeFile)\n",
    "\n",
    "    countsFile = os.path.join(newPath,'total_counts.csv')\n",
    "    aggregate_counts_data(countsFile,oldPath,stats_rows=True)\n",
    "    print('\\tAggregate counts file written to {}\\n'.format(countsFile))   \n",
    "    \n",
    "### write manifest files for QIIME2\n",
    "def writeManifestFiles(newPath='./cutprimers/cutprimers_Vsorted',\n",
    "                       manifestPath='/data/NRTS_share/ion_torrent_qiime2_methods/'):\n",
    "    \n",
    "    def getFastqFiles(path):\n",
    "        files = []\n",
    "        v_regions = set()\n",
    "        for root, dirs, filenames in os.walk(path):\n",
    "            for file in filenames:\n",
    "                if file[-6:].lower() == '.fastq' or file[-9:].lower() == '.fastq.gz':\n",
    "                    files.append(os.path.join(root,file))\n",
    "                    v_regions.add(root.split(os.sep)[-1])\n",
    "        return [os.path.normpath(f) for f in files],v_regions\n",
    "\n",
    "    ### write manifest files\n",
    "    def writeRegionManifest(writeFile,filesNew,v,sep='\\t'):  #tab:'\\t' \n",
    "        # writeFile has two columns: first is the sample ID, next is the absolute path\n",
    "        # of the file on the biowulf cluster\n",
    "        files = [f for f in filesNew if v + '.' in f]\n",
    "        fp = [(manifestPath+f).replace(os.sep,'/') for f in files]\n",
    "        sID = [os.path.basename(f).split('.')[0] for f in files]    \n",
    "        manifest = pd.DataFrame({'sample-id': sID, 'absolute-filepath': fp})\n",
    "        manifest.to_csv(writeFile,index=False,sep=sep)\n",
    "        print('\\tManifest file written to {}'.format(manifestFile))\n",
    "    \n",
    "    ## execute functions\n",
    "    filesNew,v_regions = getFastqFiles(newPath)\n",
    "    # print(v_regions)\n",
    "    for v in sorted(v_regions):\n",
    "        manifestFile = os.path.join(newPath,'manifest_{}.txt'.format(v))\n",
    "        writeRegionManifest(manifestFile,filesNew,v)\n",
    "    print()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 0. CHECK METADATA RENAME ID FILES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# IMPORTANT: Set path to rename_ids metadata information in each project and create the first \"v2_ids.tsv\" (this exact name) each time.\n",
    "cd ~/Escritorio/bioinfo_hematology_unit/tfm_david/metadata\n",
    "\n",
    "# We have an V2 file wich maps samples IDs generated by MetagenomicsPP pipeline (sample_name_V2) with the same sample name whitout the V region ending (_V2),\n",
    "# which is necessary to work with q2-sidle.\n",
    "\n",
    "# Duplicate the files for the rest V regions.\n",
    "\n",
    "# V3\n",
    "sed -ie 's/V2/V3/g' v2_ids.tsv\n",
    "\n",
    "# V4\n",
    "sed -ie 's/V2/V4/g' v2_ids.tsve\n",
    "\n",
    "# V67\n",
    "sed -ie 's/V2/V67/g' v2_ids.tsvee\n",
    "\n",
    "# V8\n",
    "sed -ie 's/V2/V8/g' v2_ids.tsveee\n",
    "\n",
    "# V9\n",
    "sed -ie 's/V2/V9/g' v2_ids.tsveeee\n",
    "\n",
    "# Change file names to make them usable.\n",
    "mv v2_ids.tsv v3_ids.tsv\n",
    "mv v2_ids.tsve v4_ids.tsv\n",
    "mv v2_ids.tsvee v67_ids.tsv\n",
    "mv v2_ids.tsveee v8_ids.tsv\n",
    "mv v2_ids.tsveeee v9_ids.tsv\n",
    "mv v2_ids.tsveeeee v2_ids.tsv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. SPLIT FASTQ FILES BY V REGION IN EACH SAMPLE.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# This is developed using a plugin from Thermofisher called MetagenomicsPP. It generates a new folder with a number of subfolder equal to the name of samples (fastq). In each \n",
    "# subfolder, it generetes an independent fastq file for each V region, another fastq file for non-adapters sequeces and summary statistics.\n",
    "\n",
    "# Note: with cutprimers, both the forward and reverse reads are combined into one \"V\" region file, and the reverse are inverse complemented so all are intepreted as forward reads.\n",
    "\n",
    "#! cd ~/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/\n",
    "\n",
    "sh ~/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/MetagenomicsPP/mgpp.sh -m ~/Escritorio/bioinfo_hematology_unit/tfm_david/data/manifest_emumyc.txt -o ~/Escritorio/bioinfo_hematology_unit/tfm_david/data/mgpp_results\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. REORGANIZE FASTQ FILES BY V REGION\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reorg_metagenomics_files(oldPath = '/home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/mgpp_results',\n",
    "                         newPath = '/home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/fastq_Vsorted')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3. CREATE A MANIFEST FILE FROM THE SUB-V FOLDERS IN QIIME2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# manifestPath = \"\": To save manifest.txt in the same folder as fastq files.\n",
    "writeManifestFiles(newPath = \"/home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/fastq_Vsorted\",\n",
    "                   manifestPath = \"\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4. IMPORT SEPARATED FASTQ FILES INTO QIIME2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input fomart is \"SingleEndFastqManifestPhred33V2\" because reverse sequences were inverse complemented\n",
    "\n",
    "# IMPORTANT: IT WILL BE NEEDED TO DISGUISH BETWEEN RUNS IN THE FUTURE. FOR NOW, WE ONLY HAVE ONE RUN.\n",
    "\n",
    "# V2 data\n",
    "! qiime tools import --type \"SampleData[SequencesWithQuality]\" \\\n",
    "    --input-format SingleEndFastqManifestPhred33V2 \\\n",
    "    --input-path /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/fastq_Vsorted/manifest_V2.txt \\\n",
    "    --output-path /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/demux_seqs/demux_seqs_V2.qza\n",
    "\n",
    "# V3 data\n",
    "! qiime tools import --type \"SampleData[SequencesWithQuality]\" \\\n",
    "    --input-format SingleEndFastqManifestPhred33V2 \\\n",
    "    --input-path /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/fastq_Vsorted/manifest_V3.txt \\\n",
    "    --output-path /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/demux_seqs/demux_seqs_V3.qza\n",
    "\n",
    "# V4 data\n",
    "! qiime tools import --type \"SampleData[SequencesWithQuality]\" \\\n",
    "    --input-format SingleEndFastqManifestPhred33V2 \\\n",
    "    --input-path /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/fastq_Vsorted/manifest_V4.txt \\\n",
    "    --output-path /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/demux_seqs/demux_seqs_V4.qza\n",
    "\n",
    "# V67 data\n",
    "! qiime tools import --type \"SampleData[SequencesWithQuality]\" \\\n",
    "    --input-format SingleEndFastqManifestPhred33V2 \\\n",
    "    --input-path /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/fastq_Vsorted/manifest_V67.txt \\\n",
    "    --output-path /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/demux_seqs/demux_seqs_V67.qza\n",
    "\n",
    "# V8 data\n",
    "! qiime tools import --type \"SampleData[SequencesWithQuality]\" \\\n",
    "    --input-format SingleEndFastqManifestPhred33V2 \\\n",
    "    --input-path /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/fastq_Vsorted/manifest_V8.txt \\\n",
    "    --output-path /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/demux_seqs/demux_seqs_V8.qza\n",
    "\n",
    "# V9 data\n",
    "! qiime tools import --type \"SampleData[SequencesWithQuality]\" \\\n",
    "    --input-format SingleEndFastqManifestPhred33V2 \\\n",
    "    --input-path /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/fastq_Vsorted/manifest_V9.txt \\\n",
    "    --output-path /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/demux_seqs/demux_seqs_V9.qza\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5. VISUALIZE QUALITY.\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 5.1 V2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "qiime demux summarize \\\n",
    "  --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/demux_seqs/demux_seqs_V2.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/demux_seqs/demux_seqs_V2.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/demux_seqs/demux_seqs_V2.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 5.2 V3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "qiime demux summarize \\\n",
    "  --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/demux_seqs/demux_seqs_V3.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/demux_seqs/demux_seqs_V3.qzv "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/demux_seqs/demux_seqs_V3.qzv "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 5.3 V4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "qiime demux summarize \\\n",
    "  --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/demux_seqs/demux_seqs_V4.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/demux_seqs/demux_seqs_V4.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/demux_seqs/demux_seqs_V4.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 5.4 V67"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "qiime demux summarize \\\n",
    "  --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/demux_seqs/demux_seqs_V67.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/demux_seqs/demux_seqs_V67.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/demux_seqs/demux_seqs_V67.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 5.5 V8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "qiime demux summarize \\\n",
    "  --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/demux_seqs/demux_seqs_V8.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/demux_seqs/demux_seqs_V8.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/demux_seqs/demux_seqs_V8.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 5.6 V9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "qiime demux summarize \\\n",
    "  --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/demux_seqs/demux_seqs_V9.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/demux_seqs/demux_seqs_V9.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/demux_seqs/demux_seqs_V9.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 6. DADA2 with denoise -pyro\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As in the previous step, this must be used with each run separately and for each V region."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 6.1 V2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# HERE, WE USED 170 AS EXAMPLE\n",
    "qiime dada2 denoise-pyro \\\n",
    "--p-trunc-len 170  \\\n",
    "--i-demultiplexed-seqs /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/demux_seqs/demux_seqs_V2.qza  \\\n",
    "--o-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_table.qza \\\n",
    "--o-representative-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_rep_seq.qza \\\n",
    "--o-denoising-stats /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_stats.qza \\\n",
    "--verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Rename sample ID from \"name_VX\" to \"name\".\n",
    "qiime feature-table rename-ids \\\n",
    "    --i-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_table.qza \\\n",
    "    --m-metadata-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/metadata/rename_ids_metadata/v2_ids.tsv \\\n",
    "    --m-metadata-column 'NewID' \\\n",
    "    --p-axis 'sample' \\\n",
    "    --o-renamed-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_table.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize stats\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_stats.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_stats.qzv\n",
    "\n",
    "# Visualize seqs\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_rep_seq.qza  \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_rep_seq.qzv\n",
    "\n",
    "# Visualize feature table\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_table.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_table.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_stats.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_rep_seq.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_table.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 6.2 V3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# HERE, WE USED 170 AS EXAMPLE\n",
    "qiime dada2 denoise-pyro \\\n",
    "--p-trunc-len 170  \\\n",
    "--i-demultiplexed-seqs /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/demux_seqs/demux_seqs_V3.qza  \\\n",
    "--o-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_table.qza \\\n",
    "--o-representative-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_rep_seq.qza \\\n",
    "--o-denoising-stats /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_stats.qza \\\n",
    "--verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Rename sample ID from \"name_VX\" to \"name\".\n",
    "qiime feature-table rename-ids \\\n",
    "    --i-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_table.qza \\\n",
    "    --m-metadata-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/metadata/rename_ids_metadata/v3_ids.tsv \\\n",
    "    --m-metadata-column 'NewID' \\\n",
    "    --p-axis 'sample' \\\n",
    "    --o-renamed-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_table.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize stats\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_stats.qza  \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_stats.qzv \n",
    "\n",
    "# Visualize seqs\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_rep_seq.qza  \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_rep_seq.qzv\n",
    "\n",
    "# Visualize feature table\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_table.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_table.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_stats.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_rep_seq.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_table.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 6.3 V4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# HERE, WE USED 170 AS EXAMPLE\n",
    "qiime dada2 denoise-pyro \\\n",
    "--p-trunc-len 170  \\\n",
    "--i-demultiplexed-seqs /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/demux_seqs/demux_seqs_V4.qza  \\\n",
    "--o-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_table.qza \\\n",
    "--o-representative-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_rep_seq.qza \\\n",
    "--o-denoising-stats /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_stats.qza \\\n",
    "--verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Rename sample ID from \"name_VX\" to \"name\".\n",
    "qiime feature-table rename-ids \\\n",
    "    --i-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_table.qza  \\\n",
    "    --m-metadata-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/metadata/rename_ids_metadata/v4_ids.tsv \\\n",
    "    --m-metadata-column 'NewID' \\\n",
    "    --p-axis 'sample' \\\n",
    "    --o-renamed-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_table.qza "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize stats\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_stats.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_stats.qzv\n",
    "\n",
    "# Visualize seqs\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_rep_seq.qza  \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_rep_seq.qzv\n",
    "\n",
    "# Visualize feature table\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_table.qza  \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_table.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_stats.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_rep_seq.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_table.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 6.4 V67"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# HERE, WE USED 170 AS EXAMPLE\n",
    "qiime dada2 denoise-pyro \\\n",
    "--p-trunc-len 170  \\\n",
    "--i-demultiplexed-seqs /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/demux_seqs/demux_seqs_V67.qza  \\\n",
    "--o-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_table.qza \\\n",
    "--o-representative-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_rep_seq.qza \\\n",
    "--o-denoising-stats /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_stats.qza \\\n",
    "--verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Rename sample ID from \"name_VX\" to \"name\".\n",
    "qiime feature-table rename-ids \\\n",
    "    --i-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_table.qza  \\\n",
    "    --m-metadata-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/metadata/rename_ids_metadata/v67_ids.tsv \\\n",
    "    --m-metadata-column 'NewID' \\\n",
    "    --p-axis 'sample' \\\n",
    "    --o-renamed-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_table.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize stats\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_stats.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_stats.qzv\n",
    "\n",
    "# Visualize feature table\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_table.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_table.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_stats.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_table.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 6.5 V8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# HERE, WE USED 170 AS EXAMPLE\n",
    "qiime dada2 denoise-pyro \\\n",
    "--p-trunc-len 170  \\\n",
    "--i-demultiplexed-seqs /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/demux_seqs/demux_seqs_V8.qza  \\\n",
    "--o-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_table.qza \\\n",
    "--o-representative-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_rep_seq.qza \\\n",
    "--o-denoising-stats /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_stats.qza \\\n",
    "--verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Rename sample ID from \"name_VX\" to \"name\".\n",
    "qiime feature-table rename-ids \\\n",
    "    --i-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_table.qza  \\\n",
    "    --m-metadata-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/metadata/rename_ids_metadata/v8_ids.tsv \\\n",
    "    --m-metadata-column 'NewID' \\\n",
    "    --p-axis 'sample' \\\n",
    "    --o-renamed-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_table.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize stats\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_stats.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_stats.qzv\n",
    "\n",
    "# Visualize feature table\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_table.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_table.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_stats.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_table.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 6.6 V9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# HERE, WE USED 170 AS EXAMPLE\n",
    "qiime dada2 denoise-pyro \\\n",
    "--p-trunc-len 170  \\\n",
    "--i-demultiplexed-seqs /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/demux_seqs/demux_seqs_V9.qza  \\\n",
    "--o-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_table.qza \\\n",
    "--o-representative-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_rep_seq.qza \\\n",
    "--o-denoising-stats /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_stats.qza \\\n",
    "--verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Rename sample ID from \"name_VX\" to \"name\".\n",
    "qiime feature-table rename-ids \\\n",
    "    --i-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_table.qza \\\n",
    "    --m-metadata-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/metadata/rename_ids_metadata/v9_ids.tsv \\\n",
    "    --m-metadata-column 'NewID' \\\n",
    "    --p-axis 'sample' \\\n",
    "    --o-renamed-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_table.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize stats\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_stats.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_stats.qzv\n",
    "\n",
    "# Visualize feature table\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_table.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_table.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_stats.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_table.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### IMPORTANT: If q2-sidle works properly, we won't need to combine feature tables and representative sequences accross V regions and we would directly obtain a feature table and a reconstructed taxonomy."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7. q2-sidle"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 7.1 Database preparation."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This step will need to be done once for each **database**, **primer pair** and **kmer lenght**, since the reference files can be reused.\n",
    "\n",
    "In our case, **primer pairs** will be the same because we'll use the same Ion Torrent kit for further experiments. Database will be SILVA 128 until the last version of SILVA (138 or upper) can be used with q2-sidle. So, **database preparation will need to be repeatead each time a new specific kmer obtained with *dada2-denoise pyro* was set**."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 7.1.1 Get SILVA 128 database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Get SILVA 128 seqs and taxonomy\n",
    "qiime rescript get-silva-data \\\n",
    "    --p-version '128' \\\n",
    "    --p-target 'SSURef_NR99' \\\n",
    "    --p-include-species-labels \\\n",
    "    --o-silva-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_rna_seqs.qza \\\n",
    "    --o-silva-taxonomy /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_tax.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Transcribe RNA sequences to DNA.\n",
    "qiime rescript reverse-transcribe \\\n",
    "    --i-rna-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_rna_seqs.qza \\\n",
    "    --o-dna-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs.qza "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize both artifacts\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/visualizations/silva_128_ssu_nr99_seqs.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "qiime metadata tabulate \\\n",
    "   --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_tax.qza \\\n",
    "   --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/visualizations/silva_128_ssu_nr99_tax.qzv "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/visualizations/silva_128_ssu_nr99_seqs.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/visualizations/silva_128_ssu_nr99_tax.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 7.1.2 “Culling” low-quality sequences with cull-seqs.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Number degenerates nucleotides: 5.\n",
    "qiime rescript cull-seqs \\\n",
    "    --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs.qza \\\n",
    "    --p-n-jobs 4 \\\n",
    "    --o-clean-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Get SILVA 128 cleaned visualization. 612 514 total sequences.\n",
    "qiime feature-table tabulate-seqs \\\n",
    " --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned.qza \\\n",
    " --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/visualizations/silva_128_ssu_nr99_seqs_cleaned.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 7.1.3  Filtering sequences by lenght and taxonomy.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# It allows us to remove low quality regions.\n",
    "qiime rescript filter-seqs-length-by-taxon \\\n",
    " --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned.qza \\\n",
    " --i-taxonomy /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_tax.qza \\\n",
    " --p-labels Archaea Bacteria Eukaryota \\\n",
    " --p-min-lens 900 1200 1400 \\\n",
    " --o-filtered-seqs /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt.qza \\\n",
    " --o-discarded-seqs /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_discarded.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Get SILVA 128 cleaned filtered (length and taxonomy) visualization. 607415 total sequences.\n",
    "qiime feature-table tabulate-seqs \\\n",
    " --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt.qza \\\n",
    " --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/visualizations/silva_128_ssu_nr99_seqs_cleaned_filt.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 7.1.4  Dereplication.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Dereplication using the default \"uniq\" approach.\n",
    " qiime rescript dereplicate \\\n",
    "  --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt.qza \\\n",
    "  --i-taxa /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_tax.qza \\\n",
    "  --p-mode 'uniq' \\\n",
    "  --p-perc-identity 0.99 \\\n",
    "  --p-threads 5 \\\n",
    "  --p-derep-prefix \\\n",
    "  --o-dereplicated-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt_derep.qza \\\n",
    "  --o-dereplicated-taxa /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_tax_derep.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Get SILVA 128 cleaned, filtered and dereplicated visualization. 392 994 sequences (36% filtered out)\n",
    "qiime feature-table tabulate-seqs \\\n",
    " --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt_derep.qza \\\n",
    " --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/visualizations/silva_128_ssu_nr99_seqs_cleaned_filt_derep.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 7.1.5  Filtering sequences by taxonomy.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Filter by taxonomy\n",
    "qiime taxa filter-seqs \\\n",
    " --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt_derep.qza \\\n",
    " --i-taxonomy /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_tax_derep.qza \\\n",
    " --p-include 'Bacteria, Archea' \\\n",
    " --p-exclude 'p__;,k__;,Chloroplast,c__Chloroplast;, Mitochondria,f__mitochondria;, Unassigned' \\\n",
    " --o-filtered-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt_derep_filttax.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Get SILVA 128 cleaned, filtered by quality, dereplicated and filterd by taxonomy visualization.  328 454 sequences.\n",
    "qiime feature-table tabulate-seqs \\\n",
    " --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt_derep_filttax.qza \\\n",
    " --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt_derep_filttax.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 7.1.6  Prepare a regional database for each primer set.\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**A. Extracting reads**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# V2\n",
    "qiime feature-classifier extract-reads \\\n",
    "    --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt_derep_filttax.qza \\\n",
    "    --p-f-primer NGCGNACGGGTGAGTAA \\\n",
    "    --p-r-primer GCTGCCTCCCGTAGGAG \\\n",
    "    --p-n-jobs 4 \\\n",
    "    --o-reads /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V2.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# V2 database visualization. 310 427 sequences.\n",
    "qiime feature-table tabulate-seqs \\\n",
    " --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V2.qza \\\n",
    " --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V2.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# V3\n",
    "qiime feature-classifier extract-reads \\\n",
    "    --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt_derep_filttax.qza \\\n",
    "    --p-f-primer CTGAGACACGGTCCANACT \\\n",
    "    --p-r-primer GTATTACCGCGGCTGCTG \\\n",
    "    --p-n-jobs 4 \\\n",
    "    --o-reads /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V3.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# V3 database visualization. 326 065 sequences.\n",
    "qiime feature-table tabulate-seqs \\\n",
    " --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V3.qza \\\n",
    " --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V3.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# V4\n",
    "qiime feature-classifier extract-reads \\\n",
    "    --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt_derep_filttax.qza \\\n",
    "    --p-f-primer CCAGCAGCCGCGGTAATA \\\n",
    "    --p-r-primer GGACTACCAGGGTATCTAATCCTGT \\\n",
    "    --p-n-jobs 4 \\\n",
    "    --o-reads /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V4.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# V4 database visualization. 326 851 sequences.\n",
    "qiime feature-table tabulate-seqs \\\n",
    " --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V4.qza \\\n",
    " --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V4.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# V67\n",
    "qiime feature-classifier extract-reads \\\n",
    "    --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt_derep_filttax.qza \\\n",
    "    --p-f-primer ACAAGCGGNGGAGCATGT \\\n",
    "    --p-r-primer GACGTCATCCCCACCTTCC \\\n",
    "    --p-n-jobs 4 \\\n",
    "    --o-reads /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V67.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# V67 database visualization. 322 578 sequeneces (211-234 bp)\n",
    "qiime feature-table tabulate-seqs \\\n",
    " --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V67.qza \\\n",
    " --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V67.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# V8\n",
    "qiime feature-classifier extract-reads \\\n",
    "    --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt_derep_filttax.qza \\\n",
    "    --p-f-primer GNTGTCGTCAGCTCGTGT \\\n",
    "    --p-r-primer CGATTACTAGCGANTCCGACTTCA \\\n",
    "    --p-n-jobs 4 \\\n",
    "    --o-reads /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V8.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# V8 database visualization. 320 902 (233-262)\n",
    "qiime feature-table tabulate-seqs \\\n",
    " --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V8.qza \\\n",
    " --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V8.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# V9\n",
    "qiime feature-classifier extract-reads \\\n",
    "    --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/silva_128_ssu_nr99_seqs_cleaned_filt_derep_filttax.qza \\\n",
    "    --p-f-primer GTTACGACTTCACCCCAGTCA \\\n",
    "    --p-r-primer GCGTCGTAGTCCGGATTGG \\\n",
    "    --p-n-jobs 4 \\\n",
    "    --o-reads /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V9.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# V8 database visualization. 73 952 (161-919)\n",
    "qiime feature-table tabulate-seqs \\\n",
    " --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V9.qza \\\n",
    " --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V9.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**B. Extracting region**"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "THIS STEP SHOULD BE REPEATED EACH TIME WE CHANGE THE TRIMMING/TRUNCATING PARAMETERS DURING DENOYSING."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# READS WERE TRIMMED AS 170 LIKE OUR SEQUENCING READS.\n",
    "\n",
    "# V2\n",
    "qiime sidle prepare-extracted-region \\\n",
    "    --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V2.qza \\\n",
    "    --p-region \"V2\" \\\n",
    "    --p-fwd-primer NGCGNACGGGTGAGTAA \\\n",
    "    --p-rev-primer GCTGCCTCCCGTAGGAG \\\n",
    "    --p-trim-length 170 \\\n",
    "    --p-n-workers 4 \\\n",
    "    --o-collapsed-kmers /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V2_170nt.qza \\\n",
    "    --o-kmer-map /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V2_170nt_map.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize V2 Kmer map.\n",
    "qiime metadata tabulate \\\n",
    " --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V2_170nt_map.qza \\\n",
    " --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V2_170nt_map.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# READS WERE TRIMMED AS 170 LIKE OUR SEQUENCING READS.\n",
    "\n",
    "# V3\n",
    "qiime sidle prepare-extracted-region \\\n",
    "    --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V3.qza \\\n",
    "    --p-region \"V3\" \\\n",
    "    --p-fwd-primer CTGAGACACGGTCCANACT \\\n",
    "    --p-rev-primer GTATTACCGCGGCTGCTG \\\n",
    "    --p-trim-length 170 \\\n",
    "    --p-n-workers 4 \\\n",
    "    --o-collapsed-kmers /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V3_170nt.qza \\\n",
    "    --o-kmer-map /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V3_170nt_map.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# READS WERE TRIMMED AS 170 LIKE OUR SEQUENCING READS.\n",
    "\n",
    "# V4\n",
    "qiime sidle prepare-extracted-region \\\n",
    "    --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V4.qza \\\n",
    "    --p-region \"V4\" \\\n",
    "    --p-fwd-primer CCAGCAGCCGCGGTAATA \\\n",
    "    --p-rev-primer GGACTACCAGGGTATCTAATCCTGT \\\n",
    "    --p-trim-length 170 \\\n",
    "    --p-n-workers 4 \\\n",
    "    --o-collapsed-kmers /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V4_170nt.qza \\\n",
    "    --o-kmer-map /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V4_170nt_map.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# READS WERE TRIMMED AS 170 LIKE OUR SEQUENCING READS.\n",
    "\n",
    "# V67\n",
    "qiime sidle prepare-extracted-region \\\n",
    "    --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V67.qza \\\n",
    "    --p-region \"V67\" \\\n",
    "    --p-fwd-primer ACAAGCGGNGGAGCATGT \\\n",
    "    --p-rev-primer GACGTCATCCCCACCTTCC \\\n",
    "    --p-trim-length 170 \\\n",
    "    --p-n-workers 4 \\\n",
    "    --o-collapsed-kmers /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V67_170nt.qza \\\n",
    "    --o-kmer-map /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V67_170nt_map.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# READS WERE TRIMMED AS 170 LIKE OUR SEQUENCING READS.\n",
    "\n",
    "# V8\n",
    "qiime sidle prepare-extracted-region \\\n",
    "    --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V8.qza \\\n",
    "    --p-region \"V8\" \\\n",
    "    --p-fwd-primer GNTGTCGTCAGCTCGTGT \\\n",
    "    --p-rev-primer CGATTACTAGCGANTCCGACTTCA \\\n",
    "    --p-trim-length 170 \\\n",
    "    --p-n-workers 4 \\\n",
    "    --o-collapsed-kmers /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V8_170nt.qza \\\n",
    "    --o-kmer-map /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V8_170nt_map.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# READS WERE TRIMMED AS 170 LIKE OUR SEQUENCING READS.\n",
    "\n",
    "# V9\n",
    "qiime sidle prepare-extracted-region \\\n",
    "    --i-sequences /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/silva_128_ssu_nr99_V9.qza \\\n",
    "    --p-region \"V9\" \\\n",
    "    --p-fwd-primer GTTACGACTTCACCCCAGTCA \\\n",
    "    --p-rev-primer GCGTCGTAGTCCGGATTGG \\\n",
    "    --p-trim-length 170 \\\n",
    "    --p-n-workers 4 \\\n",
    "    --o-collapsed-kmers /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V9_170nt.qza \\\n",
    "    --o-kmer-map /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V9_170nt_map.qza"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we have our **regional databases for each V region**  with sequences trimmed at **170 nt** as the asv sequences prepared with denoise-pyro."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 7.2  Read preparation.\n",
    "\n",
    "For this step, we have the feature tables and representative sequences (ASV) for each V region. It is important to check:\n",
    "\n",
    "- The trimmed length is the same for representative sequences and regional databases on a per V region way. For instace, if sequences comming from V4 were trimmed at 160 nt during denoise-pyro step, the sequeneces in the SILVA 128 V4 regional database must have the same lenght.\n",
    "\n",
    "- All samples must have the same names in the feature tables for each V region."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 7.3 Sequence Reconstruction."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 7.3.1 Regional alignment."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One ASV sequence can map with more than one kmer sequence from the regional database. So, the number of aligments is higher than the number of ASVs generated by denoise-pyro."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**V2 regional aligment**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# MAX_MISMATCH should be adjusted according to kmer and asv lengths.\n",
    "\n",
    "# V2\n",
    "qiime sidle align-regional-kmers \\\n",
    "    --i-kmers /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V2_170nt.qza \\\n",
    "    --i-rep-seq /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_rep_seq.qza \\\n",
    "    --p-region \"V2\" \\\n",
    "    --p-max-mismatch 3 \\\n",
    "    --p-n-workers 4 \\\n",
    "    --o-regional-alignment /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V2_aligment_map.qza \\"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize kmer aligment\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V2_aligment_map.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V2_aligment_map.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V2_aligment_map.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**V3 regional aligment**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# MAX_MISMATCH should be adjusted according to kmer and asv lengths.\n",
    "\n",
    "# V3\n",
    "qiime sidle align-regional-kmers \\\n",
    "    --i-kmers /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V3_170nt.qza \\\n",
    "    --i-rep-seq /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_rep_seq.qza \\\n",
    "    --p-region \"V3\" \\\n",
    "    --p-max-mismatch 3 \\\n",
    "    --p-n-workers 4 \\\n",
    "    --o-regional-alignment /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V3_aligment_map.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize kmer aligment\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V3_aligment_map.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V3_aligment_map.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V3_aligment_map.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**V4 regional aligment**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# MAX_MISMATCH should be adjusted according to kmer and asv lengths.\n",
    "\n",
    "# V4\n",
    "qiime sidle align-regional-kmers \\\n",
    "    --i-kmers /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V4_170nt.qza \\\n",
    "    --i-rep-seq /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_rep_seq.qza \\\n",
    "    --p-region \"V4\" \\\n",
    "    --p-max-mismatch 3 \\\n",
    "    --p-n-workers 4 \\\n",
    "    --o-regional-alignment /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V4_aligment_map.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize kmer aligment\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V4_aligment_map.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V4_aligment_map.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V4_aligment_map.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**V67 regional aligment**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# MAX_MISMATCH should be adjusted according to kmer and asv lengths.\n",
    "\n",
    "# V67\n",
    "qiime sidle align-regional-kmers \\\n",
    "    --i-kmers /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V67_170nt.qza \\\n",
    "    --i-rep-seq /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_rep_seq.qza \\\n",
    "    --p-region \"V67\" \\\n",
    "    --p-max-mismatch 3 \\\n",
    "    --p-n-workers 4 \\\n",
    "    --o-regional-alignment /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V67_aligment_map.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize kmer aligment\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V67_aligment_map.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V67_aligment_map.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V67_aligment_map.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**V8 regional aligment**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# MAX_MISMATCH should be adjusted according to kmer and asv lengths.\n",
    "\n",
    "# V8\n",
    "qiime sidle align-regional-kmers \\\n",
    "    --i-kmers /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V8_170nt.qza \\\n",
    "    --i-rep-seq /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_rep_seq.qza \\\n",
    "    --p-region \"V8\" \\\n",
    "    --p-max-mismatch 3 \\\n",
    "    --p-n-workers 4 \\\n",
    "    --o-regional-alignment /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V8_aligment_map.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize kmer aligment\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V8_aligment_map.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V8_aligment_map.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V8_aligment_map.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**V9 regional aligment**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# MAX_MISMATCH should be adjusted according to kmer and asv lengths.\n",
    "\n",
    "# V9\n",
    "qiime sidle align-regional-kmers \\\n",
    "    --i-kmers /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170/silva_128_V9_170nt.qza \\\n",
    "    --i-rep-seq /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_rep_seq.qza \\\n",
    "    --p-region \"V9\" \\\n",
    "    --p-max-mismatch 3 \\\n",
    "    --p-n-workers 4 \\\n",
    "    --o-regional-alignment /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V9_aligment_map.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Visualize kmer aligment\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V9_aligment_map.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V9_aligment_map.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/regional_alignments_sidle/V9_aligment_map.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 7.3.2 Database, Feature Table and Taxonomy Reconstruction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    from dask.distributed import Client, LocalCluster\n",
    "    cluster = LocalCluster(n_workers=2, memory_limit = \"8GB\")\n",
    "    client = Client(cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ctypes\n",
    "\n",
    "def trim_memory() -> int:\n",
    "    libc = ctypes.CDLL(\"libc.so.6\")\n",
    "    return libc.malloc_trim(0)\n",
    "\n",
    "client.run(trim_memory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster.scheduler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster.workers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "client"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "client.scheduler_info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "client.restart()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cd /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/128/artifacts/regional_databases/kmer_170\n",
    "\n",
    "qiime sidle reconstruct-database \\\n",
    "    --i-regional-alignment V2_aligment_map.qza V3_aligment_map.qza V4_aligment_map.qza V67_aligment_map.qza V8_aligment_map.qza V9_aligment_map.qza \\\n",
    "    --i-kmer-map silva_128_V2_170nt_map.qza silva_128_V3_170nt_map.qza silva_128_V4_170nt_map.qza silva_128_V67_170nt_map.qza silva_128_V8_170nt_map.qza silva_128_V9_170nt_map.qza \\\n",
    "    --p-region V2 V3 V4 V67 V8 V9 \\\n",
    "    --p-block-size 1000 \\\n",
    "    --p-client-address 'tcp://127.0.0.1:32915' \\\n",
    "    --o-database-map ./reconstructed_results/database_recons.qza \\\n",
    "    --o-database-summary ./reconstructed_results/database_recons_summ.qza \\\n",
    "    --verbose"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 7. Combining different DADA2 V regions.\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Feature tables per V region could be used to develop a comparison between V regions in terms of performance.\n",
    "\n",
    "**IMPORTANT**: If there were multiple runs, we would need to combine run specific results (feature table and representative sequences) and, then, further combination accross V regions to generate consensus results. \n",
    "\n",
    "Ej: 3 runs with 10 samples each one for the V2, V3, V4, V67, v8 and V9 regions.\n",
    "\n",
    "- **1º Combine V regions accross runs**: \n",
    "\n",
    "        V2_run1 + V2_run2 + V2_run3 = V2_merged\n",
    "\n",
    "        V3_run1 + V3_run2 + V3_run3 = V3_merged\n",
    "\n",
    "        V4_run1 + V4_run2 + V4_run3 = V4_merged\n",
    "\n",
    "        V67_run1 + V67_run2 + V67_run3 = V67_merged\n",
    "\n",
    "        V8_run1 + V8_run2 + V8_run3 = V8_merged\n",
    "\n",
    "        V9_run1 + V9_run2 + V9_run3 = V9_merged\n",
    "\n",
    "- **2º Combine accross V regions to generate consensus results (NOT RECOMMENDED)**:\n",
    "\n",
    "        V2_merged + V3_merged + V4_merged + V67_merged + V8_merged + V9_merged = consensus_results\n",
    "        "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As in this case, we only have one run, sequences were merged accros V regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "qiime feature-table merge \\\n",
    "    --i-tables /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_table.qza \\\n",
    "    --i-tables /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_table.qza \\\n",
    "    --i-tables /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_table.qza \\\n",
    "    --i-tables /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_table.qza \\\n",
    "    --i-tables /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_table.qza \\\n",
    "    --i-tables /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_table.qza \\\n",
    "    --o-merged-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_sampleid_table.qza \\\n",
    "    --p-overlap-method sum\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_sampleid_table.qza \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_sampleid_table.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_sampleid_table.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from qiime2 import Metadata\n",
    "metadata_vregions = Metadata.load(\"/home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/metadata/metadata_vregions.tsv\")\n",
    "metadata_vregions = metadata_vregions.to_dataframe()\n",
    "metadata_vregions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Group sampleid by patientid\n",
    "\n",
    "qiime feature-table group \\\n",
    "    --i-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_sampleid_table.qza \\\n",
    "    --p-axis 'sample' \\\n",
    "    --m-metadata-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/metadata/metadata_vregions.tsv \\\n",
    "    --m-metadata-column 'PatientID' \\\n",
    "    --p-mode sum \\\n",
    "    --o-grouped-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_feature_table.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_feature_table.qza\\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_feature_table.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_feature_table.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Feature Table Summary**: In this step, the main characteristics of the \"consensus\" feature table are summarized.\n",
    "\n",
    "IMPORTANT: This summary could be done with the feature tables per V region (feature tables merged for each V region) to identify how the performance differs between V regions."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Invented Metadata**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from qiime2 import Metadata\n",
    "metadata = Metadata.load(\"/home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/metadata/metadata_example.tsv\")\n",
    "metadata = metadata.to_dataframe()\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "qiime feature-table summarize \\\n",
    "    --i-table /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_feature_table.qza \\\n",
    "    --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_ft_summary.qzv \\\n",
    "    --m-sample-metadata-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/metadata/metadata_example.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_ft_summary.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 8. Combining different DADA2 sequences.\n",
    "\n",
    "In this case, the same conceptual approach is applied to sequences. As we have one run, sequences were only conmbined accros V regions. **In case of more than one run, previous combining accros runs would be needed.**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "qiime feature-table merge-seqs \\\n",
    "    --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V2/dada2_pyro/dada2_pyro_V2_rep_seq.qza \\\n",
    "    --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V3/dada2_pyro/dada2_pyro_V3_rep_seq.qza \\\n",
    "    --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V4/dada2_pyro/dada2_pyro_V4_rep_seq.qza \\\n",
    "    --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V67/dada2_pyro/dada2_pyro_V67_rep_seq.qza \\\n",
    "    --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V8/dada2_pyro/dada2_pyro_V8_rep_seq.qza \\\n",
    "    --i-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/V9/dada2_pyro/dada2_pyro_V9_rep_seq.qza \\\n",
    "    --o-merged-data /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_rep_seqs.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "qiime metadata tabulate \\\n",
    "    --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_rep_seqs.qza \\\n",
    "    --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_rep_seqs.qzv "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_rep_seqs.qzv "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 9. Classify taxonomy.\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Database OTU fasta and taxonomy files (SILVA 138) were directly downloaded from from Qiime2 as artifacts and saved in the foder used for classification files. In this case, the **full-lenght** sequences and taxonomy are used because we are going to classify sequences comming from all V regions.\n",
    "\n",
    "There is an option to extract the V region of interest, but since the sequences primers are needed to do this and Ion Torrent has them as propietary, this step can't be developed.\n",
    "\n",
    "In this case, the path to follow could be:\n",
    "\n",
    "1. To use the full-length sequences and taxonomy for each V region (feature table and sequences) to obtain results for each V region.\n",
    "\n",
    "2. To use the full-length sequences and taxonomy for each the consensus results (feature table and sequences) to obtain results for all regions together."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 9.1 Classify taxonomy for consensus results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "qiime feature-classifier classify-consensus-vsearch \\\n",
    "    --i-query /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/merged_rep_seqs.qza \\\n",
    "    --i-reference-reads /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/silva-138-99-seqs.qza \\\n",
    "    --i-reference-taxonomy /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/silva_classifier/silva-138-99-tax.qza \\\n",
    "    --p-strand 'plus' \\\n",
    "    --p-threads 2 \\\n",
    "    --o-classification /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/full_silva_taxonomy.qza \\\n",
    "    --verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "qiime metadata tabulate \\\n",
    "  --m-input-file /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/full_silva_taxonomy.qza  \\\n",
    "  --o-visualization /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/full_silva_taxonomy.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools view /home/andres/Escritorio/bioinfo_hematology_unit/myeloma_microbiome_project_2022/pipe_results/full_silva_taxonomy.qzv"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter feature table (use tesis alba ipynb)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Barplot"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 9.2 Classify taxonomy for each V region individually."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
